pkg/webhook/webhook.go

```diff
@@ -182,19 +198,113 @@ func (mw *MutatingWebhook) lookForValueFrom(env corev1.EnvVar, ns string) (*core
 			}
 			return nil, err
 		}
+
 		value := string(data[env.ValueFrom.SecretKeyRef.Key])
-		if common.HasVaultPrefix(value) || injector.HasInlineVaultDelimiters(value) {
+		if hasProviderPrefix(value, true) {
 			fromSecret := corev1.EnvVar{
 				Name:  env.Name,
 				Value: value,
 			}
 			return &fromSecret, nil
 		}
 	}
+
 	return nil, nil
 }
 
-func (mw *MutatingWebhook) newVaultClient(vaultConfig VaultConfig) (*vault.Client, error) {
+func NewMutatingWebhook(logger *slog.Logger, k8sClient kubernetes.Interface) (*MutatingWebhook, error) {
+	namespace := os.Getenv("KUBERNETES_NAMESPACE") // only for kurun
+	if namespace == "" {
+		namespaceBytes, err := os.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+		if err != nil {
+			return nil, errors.Wrap(err, "error reading k8s namespace")
+		}
+		namespace = string(namespaceBytes)
+	}
+
+	return &MutatingWebhook{
+		k8sClient: k8sClient,
+		namespace: namespace,
+		registry:  NewRegistry(),
+		logger:    logger,
+	}, nil
+}
+
+func ErrorLoggerMutator(mutator mutating.MutatorFunc, logger log.Logger) mutating.MutatorFunc {
+	return func(ctx context.Context, ar *model.AdmissionReview, obj metav1.Object) (result *mutating.MutatorResult, err error) {
+		r, err := mutator(ctx, ar, obj)
+		if err != nil {
+			logger.WithCtxValues(ctx).WithValues(log.Kv{
+				"error": err,
+			}).Errorf("Admission review request failed")
+		}
+		return r, err
+	}
+}
+
+// parseProviderConfig parses all provider configs that was declared in the webhook annotation
+func parseProviderConfig(obj metav1.Object, ar *model.AdmissionReview, providerName string) (interface{}, error) {
+	var config interface{}
+	var err error
+	switch providerName {
+	case vaultprov.ProviderName:
```

Comment:
providers can implement config and object constructors (similar to what we have on secret-init), for example
```go
// bao provider

func LoadConfig(...) (*Config, error) {
	// loads bao config
}


func New(config, ...) (Mutator, error) {
	// creates bao provider that implements Mutator/Provider.
	// replaces the local one defined here via newBaoClient
}
```

note that if you need more functionality exposed on the provider level, you can have Provider type along the Mutator, e.g.

```go
type Mutator interface {}
type Provider interface {
	Mutator

	// additional methods such as checking the prefixes...
}
```